api/src/main/java/org/openmrs/obs/handler/AbstractHandler.java

```diff
@@ -46,112 +59,144 @@ public AbstractHandler() {
 		nf.setMinimumIntegerDigits(2);
 	}
 	
+	public AbstractHandler(AdministrationService adminService, StorageService storageService) {
+		this();
+		this.adminService = adminService;
+		this.storageService = storageService;
+	}
+
 	/**
-	 * Returns a {@link File} for the given obs complex data to be written to. The output file
-	 * location is determined off of the {@link OpenmrsConstants#GLOBAL_PROPERTY_COMPLEX_OBS_DIR}
-	 * and the file name is determined off the current obs.getComplexData().getTitle().
-	 * 
-	 * @param obs the Obs with a non-null complex data on it
-	 * @return File that the complex data should be written to
+	 * @return obs dir
+	 * @since 2.8.0
 	 */
-	public File getOutputFileToWrite(Obs obs) throws IOException {
-		String title = obs.getComplexData().getTitle();
-		String titleWithoutExtension = FilenameUtils.removeExtension(title);
-		String extension = "." + StringUtils.defaultIfEmpty(FilenameUtils.getExtension(title), "dat");
-		String uuid = obs.getUuid();
-		String filename;
-		
-		if (StringUtils.isNotBlank(titleWithoutExtension)) {
-			filename = titleWithoutExtension + "_" + uuid + extension;
-		} else {
-			filename = uuid + extension;
-		}
-		
-		File dir = OpenmrsUtil.getDirectoryInApplicationDataDirectory(
-		    Context.getAdministrationService().getGlobalProperty(OpenmrsConstants.GLOBAL_PROPERTY_COMPLEX_OBS_DIR));
-		return new File(dir, filename);
+	public String getObsDir() {
+		return adminService.getGlobalProperty(OpenmrsConstants.GLOBAL_PROPERTY_COMPLEX_OBS_DIR);
 	}
-	
+
 	/**
-	 * Get the extension for a given filename if it exists, else return the filename. If there is no
-	 * filename in the input string, "raw" is returned. 
-	 * 
-	 * If given "asdf.jpg", will return "jpg".
-	 * If given "asdf", will return "asdf". 
-	 * If given "" or "a/b/c/" will return "raw".
-	 * 
-	 * @param filename
-	 * @return the part after the period in the given filename, the filename, or "raw"
-	 * @deprecated since 2.1.3 use {@link org.apache.commons.io.FilenameUtils#getExtension(String)}
-	 *             instead.
+	 * @see org.openmrs.obs.ComplexObsHandler#getObs(Obs, String)
 	 */
-	@Deprecated
-	public String getExtension(String filename) {
-		String result = FilenameUtils.getExtension(filename);
+	public Obs getObs(Obs obs, String view) {
+		// Deprecated
+		// Check if the file exists in the legacy location, if not use obs value complex as key
+		File file = AbstractHandler.getComplexDataFile(obs);
+		String key = file.exists() ? file.getAbsolutePath() : parseDataKey(obs);
```

Comment:
It's how we benefit from StorageService.getData(key) supporting absolute paths. We have much less deprecated code this way.